pkg/graveler/committed/merge.go

```diff
@@ -0,0 +1,438 @@
+	if err != nil {
+		return err
+	}
+	if baseValue != nil && bytes.Equal(sourceValue.Identity, baseValue.Identity) && bytes.Equal(sourceValue.Key, baseValue.Key) { // dest deleted this record
+		m.haveSource = m.source.Next()
+	} else {
+		if baseValue != nil && bytes.Equal(sourceValue.Key, baseValue.Key) { // deleted by dest and changed by source
+			return graveler.ErrConflictFound
+		}
+		// source added this record
+		err := m.writeRecord(sourceValue)
+		if err != nil {
+			return fmt.Errorf("write source record: %w", err)
+		}
+		m.haveSource = m.source.Next()
+	}
+	return nil
+}
+
+// handleAll handles the case where only one Iterator from source or dest remains
+func (m *merger) handleAll(iter Iterator) error {
+	for {
+		select {
+		case <-m.ctx.Done():
+			return m.ctx.Err()
+		default:
+		}
+		iterValue, iterRange := iter.Value()
+		if iterValue == nil {
+			baseRange, err := m.moveBaseToGERange(graveler.Key(iterRange.MinKey))
+			if err != nil {
+				return fmt.Errorf("base range GE: %w", err)
+			}
+			if baseRange == nil || baseRange.ID == iterRange.ID {
+				if baseRange == nil {
+					if err := m.writeRange(iterRange); err != nil {
+						return err
+					}
+				}
+				if !iter.NextRange() {
+					break
+				}
+			} else {
+				if bytes.Compare(iterRange.MaxKey, baseRange.MinKey) >= 0 { // need to enter this range
+					if !iter.Next() {
+						break
+					}
+				} else {
+					if err := m.writeRange(iterRange); err != nil {
+						return err
+					}
+					if !iter.NextRange() {
+						break
+					}
+				}
+			}
+		} else {
+			baseValue, err := m.moveBaseToGEKey(iterValue.Key)
+			if err != nil {
+				return fmt.Errorf("base value GE: %w", err)
+			}
+			if baseValue == nil || !bytes.Equal(baseValue.Identity, iterValue.Identity) {
+				if baseValue != nil && bytes.Equal(baseValue.Key, iterValue.Key) {
+					return graveler.ErrConflictFound
+				}
+				if err := m.writeRecord(iterValue); err != nil {
+					return err
+				}
+			}
+			if !iter.Next() {
+				break
+			}
+		}
+	}
+	return iter.Err()
+}
+
+// handleBothRanges handles the case where both source and dest iterators are at the header of a range
+func (m *merger) handleBothRanges(sourceRange *Range, destRange *Range) error {
+	switch {
+	case sourceRange.ID == destRange.ID: // range hasn't changed or both added the same range
+		err := m.writeRange(sourceRange)
+		if err != nil {
+			return err
+		}
+		m.haveSource = m.source.NextRange()
+		m.haveDest = m.dest.NextRange()
+
+	case bytes.Compare(sourceRange.MaxKey, destRange.MinKey) < 0: // source before dest
+		baseRange, err := m.moveBaseToGERange(graveler.Key(sourceRange.MinKey))
+		if err != nil {
+			return fmt.Errorf("base range GE: %w", err)
+		}
+		if baseRange != nil && sourceRange.ID == baseRange.ID { // dest deleted this range
+			m.haveSource = m.source.NextRange()
+			return nil
+		}
+		if baseRange == nil || destRange.ID == baseRange.ID { // source added this range
```

Comment:
The check `destRange.ID == baseRange.ID` is irrelevant, and causes an optimization bug. At this point, `destRange` starts after `sourceRange` ends. Whether or not `baseRange` is the same range is not interesting. Only thing that matters is if baseRange intersects with sourceRange or not.